### ISSUE TITLE
Spike: prototype file for Cython compatibility with small Chapel program

### ISSUE THREAD

**Issue by lydia-duncan** [chapel-lang/chapel#9239]
Create a mock file to bridge the gap between Cython and Chapel for:

- [x] a basic hello world function
- [x] a function that takes an int and returns a different int

The intention is to be sure I understand what we would need to generate based on our generated C code.

Bonus:
- [ ] a small function that operates in parallel
- [ ] a class

**Comment by lydia-duncan** [comment]
With `CHPL_LIBMODE=shared`, quickstart settings, `ARCHFLAGS="-arch i386 -arch x86_64"`, I can compile a simple Chapel program and call it from Python 3 using Cython on my Mac.

The chapel library:
[testing.h](https://github.com/lydia-duncan/chapel/blob/pythonBranchDONOTMERGE/test/compflags/lydia/library/testing.h)
[testing.chpl](https://github.com/lydia-duncan/chapel/blob/pythonBranchDONOTMERGE/test/compflags/lydia/library/testing.chpl)

Compiled using:
```
chpl --library testing.chpl -o libtesting-i386 --ccflags "-arch i386"
chpl --library testing.chpl -o libtesting-x86_64 --ccflags "-arch x86_64"
lipo -create libtesting-i386.a libtesting-x86_64.a -output libtesting.a
```

The Cython files:
[testing.pxd](https://github.com/lydia-duncan/chapel/blob/pythonBranchDONOTMERGE/test/compflags/lydia/library/testing.pxd)
[use_testing_in_python.pyx](https://github.com/lydia-duncan/chapel/blob/pythonBranchDONOTMERGE/test/compflags/lydia/library/use_testing_in_python.pyx)
[use_testing_in_python.py](https://github.com/lydia-duncan/chapel/blob/pythonBranchDONOTMERGE/test/compflags/lydia/library/use_testing_in_python.py)

Compiled using:
```
CFLAGS="-I. `$CHPL_HOME/util/config/compileline --cflags` `$CHPL_HOME/util/config/compileline --includes-and-defines`" \
LDFLAGS="-L. `$CHPL_HOME/util/config/compileline --libraries`" \
python3 use_testing_in_python.py build_ext -i
```

Run using:
```
python3
>>> import use_testing_in_python
>>> use_testing_in_python.foo_wrapper()
```

Next steps:
- [ ] investigate why the interpreter exits early.  My suspicion is the chpl_library_finalize() call
- [ ] see if I can avoid CHPL_LIBMODE, or use the regular settings (out of curiosity)
- [ ] start looking at the other codes I listed in the description

**Comment by bradcray** [comment]
What causes the -arch flags to be required?

**Comment by lydia-duncan** [comment]
When I was compiling with Cython, it would either ask for one or the other architecture, depending on which one I sent it.  This led me to believe that both were required and when I provided both, it stopped giving me the "ignoring this file" message (though there are still a slew of warnings about compiling for a later version of my operating system, which is just weird but seemingly harmless).

Interestingly, though I got the same architecture warnings for libchpl.a, it seems to find `chpl_library_init` and `chpl_library_finalize` just fine, so there might have been something else going wrong.  I'm hoping to simplify the steps down to just what worked, now that I know it is possible at all.

**Comment by lydia-duncan** [comment]
Update: It was a different problem with my set up.  A clean check out of the code as written now will work with default settings, no `CHPL_LIBMODE` set, and no `-arch` flags.  It still generates a ton of warnings, including that "ignoring file" warning I mentioned earlier - for quality of life, it would be nice to find a way to suppress those warnings (perhaps Cython has a way to hide them?)

**Comment by lydia-duncan** [comment]
Update 2: my suspicion about why the interpreter was exiting early was correct - the `chpl_library_finalize()` call does cause the termination.  `chpl_library_finalize()` will probably need to adjust so that python programs can shut the library down once the program no longer needs Chapel, but this doesn't seem like a high priority.

Here is the updated
[use_testing_in_python.pyx](https://github.com/lydia-duncan/chapel/blob/pythonBranchDONOTMERGE/test/compflags/lydia/library/use_testing_in_python.pyx), which separates the calls so that foo can be called multiple times.

**Comment by bradcray** [comment]
I agree that chpl_library_finalize() probably shouldn't cause the program to exit().  I think what's needed here is a refactor to take all of the code done in chpl_exit_common() that shuts down tasking, communication, memory, etc. and have chpl_library_finalize() call those routines without calling the exit routine itself.  Under such a refactor, ideally, chpl_exit_common() could call chpl_library_finalize() as well.  We could also consider supporting a chpl_library_exit() call that would call finalize (if it had not been) and would also exit if we thought there were use cases that would prefer to have Chapel cause the program to exit overall.

**Comment by ronawho** [comment]
Just to confirm, our shutdown code does currently call `exit()`, and we should figure out some way to separate that out from `chpl_library_finalize()`. That's easy to do for comm=none, but much trickier for gasnet (and maybe ugni.)  There is no `gasnet_shutdown()` that I know of, there is only `gasnet_exit()` which will force program termination. So far as I know there is no way to shutdown gasnet clients without exiting the entire program.

**Comment by lydia-duncan** [comment]
Thanks for the information!  I'm going to split this out into a separate issue, let's continue the discussion there.

**Comment by lydia-duncan** [comment]
I created issue #9290

**Comment by lydia-duncan** [comment]
Update 3: I was able to successfully call a Chapel function that takes and returns an integer.  See updates in [this commit](https://github.com/lydia-duncan/chapel/commit/1f41c7fbbc690b1e9da233bb7097aa234e84a18a)

**Comment by lydia-duncan** [comment]
Wrote a really dumb "parallel" program.  Need to write an interesting one to get a feel for how much Python steps on our parallelism.  Dumb version is in [this commit](https://github.com/chapel-lang/chapel/commit/1dbe9ff8c6190ae78a0c9ec1a68052361bc9f9cf)

**Comment by lydia-duncan** [comment]
Tried to write a threadring program and it does not complete, so I must be doing something wrong here.

**Comment by lydia-duncan** [comment]
Since parallelism and classes were bonus, closing and transitioning the parallelism over to #9308

**Comment by lydia-duncan** [comment]
I'm a moron, see [this comment](https://github.com/chapel-lang/chapel/issues/9308#issuecomment-383746169)

**Comment by lydia-duncan** [comment]
Interestingly enough, both the Python linked version and the C linked version of threadring execute a good 11 seconds faster than the pure Chapel version (9 seconds if I compile the library file without --fast).  Not sure if that's worth looking into but it is strange.

**Comment by lydia-duncan** [comment]
We can't export classes or records.  I believe this blocks us from using them with Python until that support is added.
